compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.td

```diff
@@ -176,27 +170,53 @@ class IREEGPU_MmaEnumAttr<EnumAttrInfo enumInfo, string name = "">
 def IREEGPU_MMAIntrinsicAttr
   : IREEGPU_MmaEnumAttr<IREEGPU_MMAIntrinsic, "mma_intrinsic">;
 
-def IREEGPU_MMAAttr : IREEGPU_MmaVectorLayoutAttr<"MMA", "MMAIntrinsicAttr"> {
+def IREEGPU_MMAAttr : AttrDef<IREEGPU_Dialect, "MMA", [
+  DeclareAttrInterfaceMethods<IREEGPU_MmaInterfaceAttr, [
+    "getABCElementTypes",
+    "getABCVectorTypes",
+    "getMNKShape",
+    "getSubgroupSize",
+    "getMmaScope",
+    "buildMmaOperation",
+    "populateOperandOffsetsSizesStrides",
+  ]>
+]> {
   let mnemonic = "mma_layout";
   let cppNamespace = "::mlir::iree_compiler::IREE::GPU";
 
-  let description = !strconcat(baseDescription, [{
-    This mma variant describes configurations for MMA ops. The |intrinsic|
-    field specifies which particular MMA intrinsic this refers to, with each
-    intrinsic implicating a specific MNK shape and operand types.
+  let description = [{
+    Attribute describing a particular shape of matrix-multiply and accumulate
+    instruction. Abstractly, all attributes of this type represent the following
+    unit of arithmetic for matrices A, B, and C.
+
+    ```
+      C += A x B
+    ```
+
+    The |intrinsic| field specifies which particular MMA intrinsic this refers
+    to, with each intrinsic implicating a specific MNK shape and operand types.
     See IREEGPUEnums.td for the definition of the intrinsics.
-  }]);
+  }];
 
-  let hasCustomAssemblyFormat = 1;
+  let parameters = (ins
+    EnumParameter<IREEGPU_MMAIntrinsic>:$intrinsic
+  );
 
-  let skipDefaultBuilders = 1;
-  let builders = [
-    AttrBuilder<(ins "MMAIntrinsic":$intrinsic)>
-  ];
+  let assemblyFormat = "`<` params `>`";
 
   let extraClassDeclaration = [{
     int64_t getBlockSize() const;
 
+    int64_t getMSize() const {
+      return std::get<0>(getMNKShape());
+    }
+    int64_t getNSize() const {
+      return std::get<1>(getMNKShape());
+    }
+    int64_t getKSize() const {
+      return std::get<2>(getMNKShape());
```

Comment:
Can these be freestanding functions rather than members? I always push back a little against members in any class, and doubly so when it's `extraClassDeclaration`.  But don't want to cause that to grow much more time-consuming for you.

Reply:
Works for me. I actually wanted to remove these getters but they were still used in a few places so I just kept the API intact.